DESIGN.md

```diff
@@ -0,0 +1,400 @@
+func Unmarshal(b []byte) (o interface{}, err error)
+func Verify(digest []byte, signatures [][]byte, verifiers []ByteVerifier) (err error)
+type Algorithm
+    func GetAlg(h *Headers) (alg *Algorithm, err error)
+type ByteSigner
+type ByteVerifier
+type Headers
+    func (h *Headers) Decode(o []interface{}) (err error)
+    func (h *Headers) DecodeProtected(o interface{}) (err error)
+    func (h *Headers) DecodeUnprotected(o interface{}) (err error)
+    func (h *Headers) EncodeProtected() (bstr []byte)
+    func (h *Headers) EncodeUnprotected() (encoded map[interface{}]interface{})
+type KeyType
+type RSAOptions
+type Sign1Message
+    func NewSign1Message() *Sign1Message
+    func (message *Sign1Message) MarshalCBOR() ([]byte, error)
+    func (m *Sign1Message) SigStructure(external []byte) ([]byte, error)
+    func (m *Sign1Message) Sign(rand io.Reader, external []byte, signer Signer) (err error)
+    func (message *Sign1Message) UnmarshalCBOR(data []byte) (err error)
+    func (m Sign1Message) Verify(external []byte, verifier Verifier) (err error)
+type SignMessage
+    func NewSignMessage() *SignMessage
+    func (m *SignMessage) AddSignature(s *Signature)
+    func (message *SignMessage) MarshalCBOR() ([]byte, error)
+    func (m *SignMessage) SigStructure(external []byte, signature *Signature) (ToBeSigned []byte, err error)
+    func (m *SignMessage) Sign(rand io.Reader, external []byte, signers []Signer) (err error)
+    func (message *SignMessage) UnmarshalCBOR(data []byte) (err error)
+    func (m *SignMessage) Verify(external []byte, verifiers []Verifier) (err error)
+type Signature
+    func NewSignature() (s *Signature)
+    func (s *Signature) Decode(o interface{})
+    func (s *Signature) Equal(other *Signature) bool
+type Signer
+    func NewSigner(alg *Algorithm, options interface{}) (signer *Signer, err error)
+    func NewSignerFromKey(alg *Algorithm, privateKey crypto.PrivateKey) (signer *Signer, err error)
+    func (s Signer) GetAlg() *Algorithm
+    func (s *Signer) Public() (publicKey crypto.PublicKey)
+    func (s *Signer) Sign(rand io.Reader, digest []byte) (signature []byte, err error)
+    func (s *Signer) Verifier() (verifier *Verifier)
+type Verifier
+    func (v *Verifier) Verify(digest []byte, signature []byte) (err error)
+```
+
+Here are the proposed APIs:
+
+```go
+const (
+    HeaderLabelAlgorithm         = 1
+    HeaderLabelCritical          = 2
+    HeaderLabelContentType       = 3
+    HeaderLabelKeyID             = 4
+    HeaderLabelCounterSignature  = 7
+    HeaderLabelCounterSignature0 = 9
+    HeaderLabelX5Bag             = 32
+    HeaderLabelX5Chain           = 33
+    HeaderLabelX5T               = 34
+    HeaderLabelX5U               = 35
+)
+
+const (
+    CBORTagSignMessage  = 98
+    CBORTagSign1Message = 18
+)
+
+const (
+    ContextSignature  = "Signature"
+    ContextSignature1 = "Signature1"
+)
+
+var (
+    AlgorithmPS256   *Algorithm

```

Comment:
Having non-const (aka mutable) global variables is a recipe for trouble, as any imported package can modify them (intentionally or unintentionally), altering the behavior of the whole applications in unforeseen ways.

What if we expose functions that return new Algorithm instances every time instead of defining those global variables. Example:

```go
func PS256() Algorithm {
    return Algorithm{
        Name: "PS256",
        Value: -37,
        HashFunc: crypto.SHA256,
    }
}
```

Reply:
This follows the same pattern as the [crypto/elliptic](https://pkg.go.dev/crypto/elliptic#Curve) package. Since `go-cose` is a security intended library, we can trade efficient for security.

Reply:
Note for when we implement this: I believe it is better to define algorithms following the [crypto.Hash](https://pkg.go.dev/crypto#Hash) pattern, where every algorithm is just a const integer whose properties are retrieved using designated functions.

Reply:
There is a drawback of the `crypto.Hash` since it does not have the crypto agility of supporting other algorithm. There might be some algorithms in private use or still under the IETF review.